src/encode/n_hints.ml

```diff
@@ -0,0 +1,103 @@
+(*
+ * encode/hints.ml --- add generic with-patterns on quantified forms
+ *
+ *
+ * Copyright (C) 2022  INRIA and Microsoft Corporation
+ *)
+
+open Ext
+open Property
+open Expr.T
+
+let with_id = "With"
+
+(** Spli [hs] into [lhs] and [rhs] where [lhs] ends with
+    the declaration of a "With(X)" predicate
+*)
+let search_with hs =
+  let rec spin lhs rhs =
+    match Deque.front rhs with
+    | None -> None
+    | Some (h, rhs) ->
+        begin match h.core with
+        (* FIXME change 'User' to 'Builtin'? *)
+        | Defn ({ core = Operator (nm, {core = Lambda ([_],_)}) }, User, _, _)
+            when nm.core = with_id ->
+            Some (Deque.snoc lhs h, rhs)
+        | _ ->
+            spin (Deque.snoc lhs h) rhs
+        end
+  in
+  spin Deque.empty hs
+
+let pol b = function
+  | Forall -> not b
+  | Exists -> b
+
+(** Make the single sequent pattern "With(x1), .., With(xn)".
+    It is assumed xs are the last variables in the context,
+    so xi is just the index (n - i + 1)
+*)
+let mk_patterns k xs =
+  let n = List.length xs in
+  let withs = List.mapi begin fun i v ->
+    (* NOTE use type info on v? *)
+    Apply (Ix k %% [], [ Ix (n - i) %% [] ]) %% []
+  end xs in
+  [ withs ]
+
+(** Effectively add a pattern "With(x)" for every bound variable "x".
+    Only negative binders that require instanciation are affected.
+*)
+let visitor = object (self : 'self)
+  inherit [bool] Expr.Visit.map as super
+
+  method expr (b, cx as scx) oe =
+    match oe.core with
+    | Apply ({ core = Internal Builtin.Neg } as op, [ e ]) ->
+        let e = self#expr (not b, cx) e in
+        Apply (op, [ e ]) @@ oe
+    | Apply ({ core = Internal Builtin.Implies } as op, [ e ; f ]) ->
+        let e = self#expr (not b, cx) e in
+        let f = self#expr (b, cx) f in
+        Apply (op, [ e ; f ]) @@ oe
+    | If (e, f, g) ->
+        let e = self#expr (not b, cx) e in
+        let f = self#expr (b, cx) f in
+        let g = self#expr (b, cx) g in
+        If (e, f, g) @@ oe
```

Comment:
I did not think polarities through when I did this, so there is a problem indeed. The same happens with the equivalence connective (each member is both left and right of an implication). I want to remove the file n_hints (see other comment above), but since I use polarities in n_rewrite and t_synth I will fix it there.